src/Sparrow/Json/AsyncBlittableJsonTextWriter.cs

```diff
@@ -8,94 +8,141 @@ namespace Sparrow.Json
-            var bytesCount = bytes.Count;
-            if (bytesCount == 0)
-                return 0;
-            await _outputStream.WriteAsync(bytes.Array, bytes.Offset, bytesCount, _cancellationToken).ConfigureAwait(false);
-            innerStream.SetLength(0);
-            return bytesCount;
+            _inner = (MemoryStream)_stream;
         }
 
         public async ValueTask WriteStreamAsync(Stream stream, CancellationToken token = default)
         {
-            await FlushAsync(token).ConfigureAwait(false);
+            var unmanagedMemory = _pinnedBuffer.Memory;
+
+            if (_pos != 0)
+            {
+                _inner.Write(unmanagedMemory.Memory.Span.Slice(0, _pos));
+                _pos = 0;
+                _started = true;
+            }
 
             while (true)
             {
-                _pos = await stream.ReadAsync(_pinnedBuffer.Memory.Memory, token).ConfigureAwait(false);
-                if (_pos == 0)
+                var read = await stream.ReadAsync(unmanagedMemory.Memory, token).ConfigureAwait(false);
+                if (read == 0)
                     break;
 
-                await FlushAsync(token).ConfigureAwait(false);
+                _inner.Write(unmanagedMemory.Memory.Span.Slice(0, read));
+                _started = true;
+
+                await MaybeFlushAsync(token).ConfigureAwait(false);
             }
+
+            await FlushAsync(token).ConfigureAwait(false);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public ValueTask<int> MaybeFlushAsync(CancellationToken token = default)
         {
-            var innerStream = _stream as MemoryStream;
-            if (innerStream == null)
-                ThrowInvalidTypeException(_stream?.GetType());
-            if (innerStream.Length * 2 <= innerStream.Capacity)
+            if (_inner.Length * 2 <= _inner.Capacity)
                 return new ValueTask<int>(0);
 
-            FlushInternal(); // this is OK, because inner stream is a MemoryStream
             return FlushAsync(token);
         }
 
         public async ValueTask<int> FlushAsync(CancellationToken token = default)
         {
-            var innerStream = _stream as MemoryStream;
-            if (innerStream == null)
-                ThrowInvalidTypeException(_stream?.GetType());
             FlushInternal();
-            innerStream.TryGetBuffer(out var bytes);
+
+            var currentStream = _inner;
+            currentStream.TryGetBuffer(out var bytes);
+
             var bytesCount = bytes.Count;
-            if (bytesCount == 0)
-                return 0;
-            await _outputStream.WriteAsync(bytes.Array, bytes.Offset, bytesCount, token).ConfigureAwait(false);
-            innerStream.SetLength(0);
+
+            if (_outputStream is MemoryStream)
+            {
+                // We know it is safe to write and flush synchronously.
+                _outputStream.Write(bytes.Array, bytes.Offset, bytesCount);
```

Comment:
Can we use `WriteAsync` and pass the cancellation token?

Reply:
No, that is a `MemoryStream`, the whole idea is that if you are using one, you can avoid the async machinery that is the cause of the problem.